### PR TITLE
Add linalg_ext.scatter operation.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Also some minor changes
- move some checks to tablegen
- add some extra verifications to the Interface.
- move the ops to use declerative assembly format.

### CODE REVIEW

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 1]
```diff
@@ -31,10 +34,28 @@ class LinalgExt_Op<string mnemonic, list<OpTrait> traits = []> :
...
-    AttrSizedOperandSegments,
-    DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
-    SingleBlockImplicitTerminator<"YieldOp">]> {
+def LinalgExt_ScatterOp : LinalgExt_Op<"scatter"> {
+  let summary = "Scatter operator";
+  let description = [{
+    Based on XLA operation semantics, generates a result which is the
+    value of the input array `operand`, with several slices (at
+    indices specified `indices`) updated with the values in `updates`
+    using `update_computation`.
```
Comment (hanhanW) [inline iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:43]: Can we add a description for region arguments? I think we can say that it must contain to block argument. The first argument is update value and the second argument is init value?

mhlo.scatter has a reverse form -- `output[O] = update_computation(output[O], updates[U])`.

I think `output[O] = update_computation(updates[U], output[O])` matches Linalg ops more. We can handle the mapping when lowering mhlo.scatter to linalg_ext ops.

**Review by hanhanW — APPROVED** [review]
Thanks!!! The assembly format looks really great! Just one nit in scatter op definition.

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 1 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: Ok, I updated the description

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 2]
```diff
@@ -31,10 +34,33 @@ class LinalgExt_Op<string mnemonic, list<OpTrait> traits = []> :
...
+  let summary = "Scatter operator";
+  let description = [{
+    Based on XLA operation semantics, takes two `inputs` (`update` and
+    `indices`) and `outputs` value (`original`). The operation updates
+    the value at the slices specified by `indices` by combining the
+    current value with the value in `updates` using the computation
+    specified in `region`. The `region` specifies a binary operation
+    of signature (T, T) -> T, where `T` is the element-type of
+    `updates` (and `original`). The first argument correspond the
+    value to be updated (i.e. from `updates), and the second the
```
Comment (hanhanW) [inline iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:47]: nit: it's not closed.

```suggestion
    value to be updated (i.e. from `updates`), and the second the
```

**Pull request merged**